### PR TITLE
Update sabre to 2.1

### PR DESCRIPTION
**Pull request by PVince81**
Depends on core's sabre update: https://github.com/owncloud/core/pull/14151

This at least lets the calendar app load.
I haven't tested whether CalDAV still works.

Also, do you have an idea what "Principals" are ?

@georgehrke

### CODE REVIEW

**Comment by PVince81** [comment]
Also I'm not 100% sure whether those lines must be removed or replaced with others.
I thought that the new Sabre was automatically mapping these to internal string properties that have all the necessary fixes.

Note that StringProperty was removed in core, see https://github.com/owncloud/core/pull/14151#discussion-diff-24604921 and https://github.com/owncloud/core/pull/14151#discussion-diff-24604937

**Comment by georgehrke** [comment]
StringProperty was introduced by tanghus. IIRC it was to fix escaping issues that incorrectly replaced , to \, etc. 
I am not sure if this was fixed in Sabre VObject, but I guess so.

**Comment by PVince81** [comment]
Might help: http://sabre.io/vobject/upgrade/
But not sure what the `$classMap` replacement is.

**Comment by PVince81** [comment]
@evert I don't see any mention of `Sabre\VObject\Property::$classMap` in the update guide here: http://sabre.io/vobject/upgrade/

Or was `$classMap` not the correct way in the first place ?

**Comment by georgehrke** [comment]
If I am not completely mistaken `$classmap` has been split up into `$propertyMap` and `$componentMap`

**Comment by georgehrke** [comment]
but nevertheless: If the escaping issue (, to \,) was fixed in upstream (which I am pretty sure about), there is no need for classmap respectively propertyMap in the calendar app anymore.

**Comment by PVince81** [comment]
Ok cool. I'll still need to figure it out for the contacts app though.

Can you help testing if CalDAV still works ?

**Comment by georgehrke** [comment]
> Can you help testing if CalDAV still works ?

Sure, will try to find some time in the next couple of days.

**Comment by PVince81** [comment]
Take your time. The core PR still needs some work :smile:

**Comment by evert** [comment]
@PVince81 , the classMap indeed went to two new properties, and they can be found in `Sabre\VObject\Component\VCalendar` and `Sabre\VObject\Component\VCard`, but as @georgehrke said, I believe you no longer need them... and if you do, it's probably wiser to try to work around it.

**Comment by evert** [comment]
Principals are users and groups in webdav terms. Any caldav or carddav user must also have a principal node that the client uses to fetch information about the user, as well as a few other things.

**Comment by PVince81** [comment]
I had a closer look and it looks like updating the calendar app will be more than just tweaking stuff.

A few things that will need to be adapted:
- [x] replace "Component::create('VCALENDAR')" with "new VCalendar()"
- [x] replace "Component::create('VEVENT') with "$vcard->createComponent('VEVENT')"
- [ ] fix date time handling everywhere (getDateType() is gone, getDateTimeZone() seems to be different too)
- [x] update Sabre backends to work with the new PropFind/PropPatch objecst
- [x] update name of sabre classes that came from core (OC_Connector_Sabre_XXX => OC\Connector\Sabre\XXX)

**Comment by georgehrke** [comment]
will take care of the changes

**Comment by PVince81** [comment]
I have a few changes locally (VCard), will push soon.
I started with the DateTime but I got confused and is messy, so I'll revert.

**Comment by georgehrke** [comment]
> I have a few changes locally (VCard), will push soon.

Ok, but VCard sounds wrong. Is there no VCalendar class?

**Comment by PVince81** [comment]
@georgehrke ideally we should try avoid using OC_VObject from core altogether, if possible.

**Comment by georgehrke** [comment]
Yes, I totally agree, but VCards are for contact-cards. There is probably a dedicated VCalendar class (in vobject)

**Comment by PVince81** [comment]
Ah sorry, I mixed them up. It's VCalendar that needs instantiating.

I pushed my change, feel free to take over now.

**Comment by georgehrke** [comment]
> I pushed my change, feel free to take over now.

Thanks, will do

**Comment by osos** [comment]
Will this automitically fix calendar invites, issue #318

**Comment by PVince81** [comment]
@georgehrke any update ? Do you need help ?

I was able to fix the contacts app so I might be able to help a bit. My only problem here is that I don't understand all these different date formats (and why they are there) and how to convert them to the new API.

**Comment by georgehrke** [comment]
Had a lot of things to do for university last 2 weeks, should find some time this week

**Comment by PVince81** [comment]
Ok cool. Please shout if you need help.

**Comment by georgehrke** [comment]
Note to myself:
- `getDateType` the way we use it can be replaced with `isFloating`
- `getDateTimeZone` with $element['TZID']

File: lib/sabre/backend.php [hunk 1]
```diff
@@ -1,5 +1,7 @@
 <?php
 
+use \Sabre\DAV\PropPatch;
+
 class OC_Connector_Sabre_CalDAV extends \Sabre\CalDAV\Backend\AbstractBackend {
```
Comment (nickvergessen) [inline lib/sabre/backend.php:5]: Please fix the namespace while on it.
See https://github.com/owncloud/calendar/pull/653 (change namespace and remove classpath hack in app.php)

**Comment by DeepDiver1975** [comment]
@georgehrke please try to finish this asap - we need to provide an update for 8.1 on the appstore.
Still 4 weeks to go - but fixing this asap does not hurt :wink:

**Comment by PVince81** [comment]
Rebased

**Comment by PVince81** [comment]
@georgehrke how about merging this first ?
At least with this PR it is possible to enable the app (no fatal error).

Then continue in separate PRs for the date management and other things ?

**Comment by PVince81** [comment]
I rechecked the TODO list from https://github.com/owncloud/calendar/pull/722#issuecomment-75223051 and it looks like I was already done with VCalendar and VEVENT. So the only remaining stuff is converting the dates (and of course extensive testing/bugfixing)

**Comment by nickvergessen** [comment]
@PVince81 mint fixing https://github.com/owncloud/calendar/pull/722#discussion_r27372630 before merging? ;)

**Comment by PVince81** [comment]
This can be merged afterwards.

**Comment by PVince81** [comment]
Ok, seems I was confused by some methods missing in the API. They're not missing but were custom methods in core's "VObject" override. Let's see how to replace these.

**Comment by PVince81** [comment]
@evert is this:

```
$vevent->DTSTART = $now;
```

the same as

```
$vevent->create('DTSTART')->setValue($now);
```

?

What happens if I call create twice on the same property, will it be duplicated ?

**Comment by PVince81** [comment]
There are too many ways to do the same thing, this is very confusing.

Trying to figure out how to properly set VEVENT properties. I had $vevent->DTSTART->setValue() is not defined, maybe I need to create the property first.

**Comment by PVince81** [comment]
Looks like the $vevent object didn't have the properties by default, this did the trick:

``` php
            $start = new DateTime($from.' '.$fromtime, $timezone);
            $end = new DateTime($to.' '.$totime, $timezone);
            $start->setTimeZone($timezone);
            $end->setTimeZone($timezone);
            $dtstart = $vcalendar->create('DTSTART');
            $vevent->DTSTART = $dtstart;
            $dtend = $vcalendar->create('DTEND');
            $vevent->DTEND = $dtend;
            $dtstart->setValue($start);
            $dtend->setValue($end);
```

**Comment by georgehrke** [comment]
Yes, we should merge this, so the fatal error is gone and you can at least use your ownCloud instance when the calendar is installed.

**Comment by PVince81** [comment]
Ok, will merge this first. I have some local changes for the dates, will raise separately.

**Comment by PVince81** [comment]
Continuing here: https://github.com/owncloud/calendar/pull/771

**Comment by evert** [comment]
Hey @PVince81, your last code snippet:

```
        $start = new DateTime($from.' '.$fromtime, $timezone);
        $end = new DateTime($to.' '.$totime, $timezone);
        $start->setTimeZone($timezone);
        $end->setTimeZone($timezone);
        $dtstart = $vcalendar->create('DTSTART');
        $vevent->DTSTART = $dtstart;
        $dtend = $vcalendar->create('DTEND');
        $vevent->DTEND = $dtend;
        $dtstart->setValue($start);
        $dtend->setValue($end);
```

Is rather dense =). This _should_ also work:

```
        $start = new DateTime($from.' '.$fromtime, $timezone);
        $end = new DateTime($to.' '.$totime, $timezone);

        $vevent->DTSTART = $start;
        $vevent->DTEND = $end;
```

If that's too magic:

```
        $start = new DateTime($from.' '.$fromtime, $timezone);
        $end = new DateTime($to.' '.$totime, $timezone);

        $vevent->add('DTSTART', $start);
        $vevent->add('DTEND', $end);
```

The difference between my first and second replacement is that the first _sets_ the DTSTART, and the latter _adds_ a property. So the latter would add a secondary `DTSTART` property if one already existed.

**Comment by georgehrke** [comment]
Applied @evert's suggestions in https://github.com/owncloud/calendar/commit/aa71f8f4b29e4c913eb58846379cecc9b50533cc

**Pull request merged**